relay/relay.go

```diff
@@ -273,11 +275,36 @@ func (r *Relay) process(parentCtx context.Context) error {
 		}
 	}()
 
+	readerRetry, err := retry.NewReaderRetry(r.cfg.ReaderRetry)
+	if err != nil {
+		return err
+	}
+
 	transformer2 := transformer.NewTransformer(parser2)
 
 	go r.doIntervalOps(parentCtx)
 
-	return r.handleEvents(parentCtx, reader2, transformer2, writer2)
+	// handles binlog events with retry mechanism.
+	// it only do the retry for some binlog reader error now.
+	for {
+		err := r.handleEvents(parentCtx, reader2, transformer2, writer2)
+		if err == nil {
+			return nil
+		} else if !readerRetry.Check(parentCtx, err) {
+			return err
+		}
+
+		r.tctx.L().Info("receive retryable error for binlog reader", log.ShortError(err))
+		err = reader2.Close() // close the previous reader
```

Comment:
can we close `reader2` twice?

Reply:
YES, we can close `Reader` any times, but it will return a `can not close` error except the first one.

In this place, we only close a `reader2` instance once. this L298 close the previous one, and in the `defer` in L260 ~ L265, it may close the first one (without retry) or the latest one (with retry).

Reply:
check `reader2 != nil` in https://github.com/pingcap/dm/pull/265/commits/c6baaabfd39e63573a1d61ccbd8e1ebb1582e380.